### PR TITLE
Adding undo_split and a test thereof (+small lint)

### PR DESCRIPTION
**Pull request by rfl-urbaniak**
`counterfactual.ops.py` now contains a definition of `undo_split`, which is tested at the bottom of `tests.indexed.test_internals.py`.

### CODE REVIEW

File: chirho/counterfactual/ops.py [hunk 1]
```diff
@@ -23,10 +25,61 @@ def split(obs: T, acts: Tuple[Intervention[T], ...], **kwargs) -> T:
     return scatter(act_values, event_dim=kwargs.get("event_dim", 0))
 
 
+@pyro.poutine.runtime.effectful(type="undo_split")
```
Comment (eb8680) [inline chirho/counterfactual/ops.py:28]: `undo_split` is just a pure function, it doesn't need to be marked `effectful`.

File: chirho/counterfactual/ops.py [hunk 2]
```diff
@@ -23,10 +25,61 @@ def split(obs: T, acts: Tuple[Intervention[T], ...], **kwargs) -> T:
...
+def undo_split(
```
Comment (eb8680) [inline chirho/counterfactual/ops.py:29]: Can you move `undo_split` to a new file `chirho/handlers/counterfactual/explanation.py` that we will use for subsequent PRs in this series?

File: tests/indexed/test_internals.py [hunk 3]
```diff
@@ -474,7 +545,25 @@ def test_cond_tensor_associate(enum_shape, batch_shape, plate_shape, event_shape
...
+    assert (
+        actual_full.shape
+        == enum_shape + batch_shape + plate_shape + event_shape
+    )
     assert actual_full.shape == actual_left.shape == actual_right.shape
     assert (actual_full == actual_left).all()
     assert (actual_left == actual_right).all()
+
+
+def test_undo_split():
```
Comment (eb8680) [inline tests/indexed/test_internals.py:557]: Can you move this test into a new file `tests/counterfactual/test_handlers_explanation.py` and parametrize it by batch and event shapes similar to other tests here?

File: chirho/counterfactual/ops.py [hunk 4]
```diff
@@ -23,10 +25,61 @@ def split(obs: T, acts: Tuple[Intervention[T], ...], **kwargs) -> T:
...
+    antecedents: Iterable[str] = None, event_dim: int = 0
+) -> Callable[[T], T]:
+    """
+    A helper function that undoes an upstream `chirho.counterfactual.ops.split`
+    operation by gathering the factual value and scattering it back into
+    two alternative cases.
+
+    :param antecedents: A list of upstream intervened sites which induced
+                        the `split` to be reversed.
+    :param event_dim: The event dimension.
```
Comment (eb8680) [inline chirho/counterfactual/ops.py:39]: nit: clarify whose event dimension this is (the value to be preempted)

File: chirho/counterfactual/ops.py [hunk 5]
```diff
@@ -23,10 +25,61 @@ def split(obs: T, acts: Tuple[Intervention[T], ...], **kwargs) -> T:
     return scatter(act_values, event_dim=kwargs.get("event_dim", 0))
 
 
+@pyro.poutine.runtime.effectful(type="undo_split")
+def undo_split(
+    antecedents: Iterable[str] = None, event_dim: int = 0
+) -> Callable[[T], T]:
+    """
+    A helper function that undoes an upstream `chirho.counterfactual.ops.split`
```
Comment (eb8680) [inline chirho/counterfactual/ops.py:33]: Explain that `undo_split` is meant to be used to create arguments to pass to `intervene`/`split`/`preempt`

**Review by eb8680 — CHANGES_REQUESTED** [review]
Looks like a great start. I left some comments, but the main thing to do is add a few more tests for `undo_split` that mix multiple `split` and `undo_split` calls in various ways and parametrize the tests with batch and event shapes. You could also add a test that passes `undo_split` to `preempt`. The stronger and more extensive the tests for `undo_split` and `preempt`, the easier it will be to get everything else in subsequent PRs working.

Also, could you revert all of the formatting changes? I can't reproduce these locally with `make lint` and `make format`, which are the source of truth for linter settings in the ChiRho codebase.

File: chirho/counterfactual/ops.py (hunk 2 above)
Comment (rfl-urbaniak) [reply chirho/counterfactual/ops.py:29]: I take it you meant `chirho/counterfactual/handlers/explanation.py` here, which matches the existing folder structure?

File: chirho/counterfactual/ops.py (hunk 1 above)
Comment (rfl-urbaniak) [reply chirho/counterfactual/ops.py]: Right. Removed.

File: tests/indexed/test_internals.py (hunk 3 above)
Comment (rfl-urbaniak) [reply tests/indexed/test_internals.py]: moved away, will parametrize, now testing combinations with other ops.

File: chirho/counterfactual/ops.py (hunk 2 above)
Comment (eb8680) [reply chirho/counterfactual/ops.py]: Sorry, yes, that's right

Comment (rfl-urbaniak) [reply chirho/counterfactual/ops.py]: done.

File: chirho/counterfactual/ops.py (hunk 4 above)
Comment (rfl-urbaniak) [reply chirho/counterfactual/ops.py]: done

File: chirho/counterfactual/ops.py (hunk 5 above)
Comment (rfl-urbaniak) [reply chirho/counterfactual/ops.py]: done

**Comment by eb8680** [comment]
I marked this PR and #250 as blocked because I'd like to hold off on merging those into `master` until after we cut the [CZI demo release](https://github.com/BasisResearch/chirho/milestone/6) on Wednesday morning so that we don't have an incomplete `explanation` module in the demo release. We can land these two PRs and the other tasks in the [causality milestone](https://github.com/BasisResearch/chirho/milestone/7) in the next release.

File: tests/indexed/test_internals.py (hunk 3 above)
Comment (rfl-urbaniak) [reply tests/indexed/test_internals.py]: done.

**Comment by eb8680** [comment]
@rfl-urbaniak can you run `make format` to fix the remaining linting error?

**Review by eb8680 — APPROVED** [review]
Should be good to go once tests pass

**Pull request merged**